[DSIP-88][Auth] Enhancing Apache DolphinScheduler with Generalized OIDC Authentication

tusaryan opened the pull request:
<!--Thanks very much for contributing to Apache DolphinScheduler, we are happy that you want to help us improve DolphinScheduler! -->

## Purpose of the pull request
This pull request introduces end-to-end support for **OIDC (OpenID Connect) authentication**, as outlined in [DSIP-88](https://github.com/apache/dolphinscheduler/issues/17171). This feature enhances Apache DolphinScheduler by allowing users to log in via configured external identity providers like Keycloak, Okta, Google, etc.

This single, comprehensive PR delivers the complete feature, including:
1.  **Core Backend:** A foundational configuration structure and the primary `OidcAuthenticator` service to handle all OIDC protocol interactions.
2.  **Frontend UI:** A seamless user interface on the login page that dynamically displays SSO buttons for each configured provider.
3.  **API & Integration Tests:** A full suite of unit and API tests to ensure the reliability and correctness of the entire OIDC flow.

By enabling this, DolphinScheduler becomes more accessible and secure for enterprise environments that rely on centralized identity management.

Closes: #17171

<!--(For example: This pull request adds checkstyle plugin).-->

## Brief change log

<!--*(for example:)*
- *Add maven-checkstyle-plugin to root pom.xml*
-->

It covers the implementation of the OIDC feature across the backend, frontend, and testing modules.

### Core Backend & API
- **Feature:** Added new configuration classes (`OidcConfigProperties`, `OidcProviderConfig`, `OidcUserConfig`) to load OIDC settings from `application.yaml`.
- **Feature:** Implemented the `OidcAuthenticator` class using the Nimbus SDK to handle OIDC discovery, token exchange, and user information retrieval.
- **Feature:** Added three new endpoints to `LoginController` to manage the OIDC flow:
    - `/oidc-providers`: Exposes the list of configured providers to the UI.
    - `/oauth2/authorization/{providerId}`: Initiates the login flow and redirects users.
    - `/login/oauth2/code/{providerId}`: Handles the callback from the IdP to create a user session.
- **Enhancement:** Updated `AuthenticationType.java` and `SecurityConfig.java` to register OIDC as a new authentication method.
- **Enhancement:** Updated `AppConfiguration.java` to exclude the new OIDC endpoints from the login interceptor, allowing unauthenticated access for the login process.

### Frontend UI
- **API Service Integration:** Added the `getOidcProviders` function and `OidcProvider` type definition in the service layer to fetch and handle provider data from the backend.
- **Login Page Logic:** Updated the `use-login.ts` hook to fetch providers on page load and to process the post-login callback by validating the `sessionId` from the URL.
- **UI Rendering and Styling:** The `index.tsx` login page now dynamically renders a styled button for each OIDC provider, complete with the provider's logo (`iconUri`) and an interactive hover effect defined in `index.module.scss`.

### API & Integration Tests
- **Feature:** Added a new API test class, `OidcLoginAPITest.java`, to validate the entire OIDC login lifecycle.
- **Feature:** Introduced a Docker-based test environment (`docker-compose.yaml` and `realm-export.json`) that automatically sets up a Keycloak instance as the OIDC provider for realistic integration testing.
- **Enhancement:** The test framework's `HttpResponse` entity was updated to capture response headers, which is essential for validating the HTTP redirects in the OIDC flow.

## Verify this pull request

<!--*(Please pick either of the following options)*-->

This change is covered by both new automated tests and can be verified manually.

- **Automated Tests:**
    - A comprehensive suite of new **unit tests** (`OidcAuthenticatorTest`, `LoginControllerOidcTest`, etc.) has been added to verify the core backend logic.
    - A new **API test suite** (`OidcLoginAPITest`) has been added to validate the full API flow against a live Keycloak instance.

- **Manual Verification Steps:**
    1. Ensure the backend is running with at least one OIDC provider configured in `application.yaml`.
    2. Start the `dolphinscheduler-ui` service.
    3. Navigate to the login page. You should see the new, styled OIDC provider button(s) below the main login form.
    4. Click on an OIDC provider button. You should be correctly redirected to the external identity provider's login page.
    5. Authenticate successfully with the external provider.
    6. After authentication, you should be redirected back to the DolphinScheduler dashboard, fully logged in.

<!--*(example:)*
- *Added dolphinscheduler-dao tests for end-to-end.*
- *Added CronUtilsTest to verify the change.*
- *Manually verified the change by testing locally.* -->

## Pull Request Notice
[Pull Request Notice](https://github.com/apache/dolphinscheduler/blob/dev/docs/docs/en/contribute/join/pull-request.md)

If your pull request contains incompatible change, you should also add it to `docs/docs/en/guide/upgrade/incompatible.md`

tusaryan committed:
Add OpenID Connect (OIDC) authentication support

Introduces OIDC as a new authentication type in AuthenticationType and SecurityConfig. Adds OidcAuthenticator and related configuration classes to support OIDC providers, user mapping, and group-based roles. Includes unit tests for OIDC configuration and authenticator logic.

tusaryan committed:
Add OpenID Connect (OIDC) authentication support

Introduces OIDC as a new authentication type in AuthenticationType and SecurityConfig. Adds OidcAuthenticator and related configuration classes to support OIDC providers, user mapping, and group-based roles. Includes unit tests for OIDC configuration and authenticator logic.

tusaryan committed:
Add OIDC authentication and API base URL config to application.yml

Introduces OIDC authentication support with provider configuration, auto-provisioning options, and admin group mapping. Adds 'base-url' and 'ui-url' properties for API and UI endpoints, and updates the database password.

tusaryan committed:
Add OIDC authentication and API base URL config to application.yml

Introduces OIDC authentication support with provider configuration, auto-provisioning options, and admin group mapping. Adds 'base-url' and 'ui-url' properties for API and UI endpoints, and updates the database password.

tusaryan committed:
Update application.yaml

tusaryan committed:
Update application.yaml

Gallardot commented:
@tusaryan Sorry for the delayed response. Thank you for your contribution.
This Pull Request is strongly related to #17341 and should be combined into a single one. Once merged, we'll review it together. Thank you.

tusaryan committed:
Add OIDC authentication support to LoginController

Introduces OIDC authentication endpoints and logic to LoginController, including provider listing, OIDC login redirection, and callback handling. Updates AppConfiguration to allow new OIDC-related paths. Adds comprehensive unit tests for OIDC authentication scenarios in LoginControllerOidcTest and extends LoginControllerTest with additional login validation cases.

tusaryan committed:
Add OIDC authentication support to LoginController

Introduces OIDC authentication endpoints and logic to LoginController, including provider listing, OIDC login redirection, and callback handling. Updates AppConfiguration to allow new OIDC-related paths. Adds comprehensive unit tests for OIDC authentication scenarios in LoginControllerOidcTest and extends LoginControllerTest with additional login validation cases.

tusaryan committed:
Add and improve unit tests for OIDC config and authenticator

tusaryan committed:
Add and improve unit tests for OIDC config and authenticator

tusaryan committed:
apply spotless formatting

tusaryan committed:
apply spotless formatting

SbloodyS committed:
Merge branch 'dev' into feature/17171-oidc-core-backend

tusaryan commented:
Can anyone help me identify or fix the error?
I've reviewed the unit test error log, but it doesn't appear to be related to the OIDC unit tests. If I'm mistaken, please let me know or point me in the right direction. Thanks!

Gallardot committed:
Merge branch 'dev' into feature/17171-oidc-core-backend

Gallardot committed:
Merge branch 'dev' into feature/17171-oidc-core-backend

SbloodyS commented:
> Can anyone help me identify or fix the error? I've reviewed the unit test error log, but it doesn't appear to be related to the OIDC unit tests. If I'm mistaken, please let me know or point me in the right direction. Thanks!

UT have been passed.

Gallardot committed:
Merge branch 'dev' into feature/17171-oidc-core-backend

tusaryan committed:
add API tests for OIDC authentication flow

This commit introduces a new API test suite for the OIDC authentication feature.
It includes:
1. A new test class `OidcLoginAPITest.java` to validate the OIDC login and callback endpoints.
2. A Page Object `OidcLoginPage.java` for interacting with OIDC API endpoints.
3. A Docker Compose setup (`docker/oidc-login`) with a pre-configured Keycloak instance to create a self-contained testing environment for OIDC.
4. Tests for provider discovery, login initiation, and various callback scenarios, including error handling and redirects.

tusaryan committed:
add API tests for OIDC authentication flow

This commit introduces a new API test suite for the OIDC authentication feature.
It includes:
1. A new test class `OidcLoginAPITest.java` to validate the OIDC login and callback endpoints.
2. A Page Object `OidcLoginPage.java` for interacting with OIDC API endpoints.
3. A Docker Compose setup (`docker/oidc-login`) with a pre-configured Keycloak instance to create a self-contained testing environment for OIDC.
4. Tests for provider discovery, login initiation, and various callback scenarios, including error handling and redirects.

tusaryan committed:
[Feature][UI]: Implement OIDC provider support on Login page

This commit introduces the complete frontend flow for OIDC authentication.

API Integration: Adds the service calls (getOidcProviders) and TypeScript types (OidcProvider) needed to communicate with the backend.
Dynamic UI: The login page now fetches and dynamically renders a styled button for each configured OIDC provider.
Redirect Handling: Implements the logic in use-login.ts to handle the callback from the OIDC provider, establishing a user session seamlessly.
Styling & others: Includes custom SCSS for the login buttons and updates localization files with new text.

tusaryan committed:
[Feature][UI]: Implement OIDC provider support on Login page

This commit introduces the complete frontend flow for OIDC authentication.

API Integration: Adds the service calls (getOidcProviders) and TypeScript types (OidcProvider) needed to communicate with the backend.
Dynamic UI: The login page now fetches and dynamically renders a styled button for each configured OIDC provider.
Redirect Handling: Implements the logic in use-login.ts to handle the callback from the OIDC provider, establishing a user session seamlessly.
Styling & others: Includes custom SCSS for the login buttons and updates localization files with new text.

tusaryan committed:
Fix: Harden OIDC security, resolve CI issues and Add iconUri support for OIDC providers

This commit addresses vulnerabilities, updates configuration examples, and resolves code quality issues flagged by CI in the OIDC implementation.

- Security:
  - Sanitize `providerId` in `OidcAuthenticator` before logging to prevent log injection.
  - Validate `providerId` in `LoginController` to fix a trust boundary violation.

- Refactor:
  - Removed unused `providerId` and `request` parameters from the OIDC callback method.

- Feature:
  - Added an `iconUri` field to `OidcProviderConfig` to support custom provider icons on the login page.

- Configuration:
  - Updated `application.yaml` to include the new `icon-uri` property in the Keycloak provider example, demonstrating the feature.
  - Improved the commented-out `issuer-uri` example for better use in containerized environments (e.g., Docker).

tusaryan committed:
Fix: Harden OIDC security, resolve CI issues and Add iconUri support for OIDC providers

This commit addresses vulnerabilities, updates configuration examples, and resolves code quality issues flagged by CI in the OIDC implementation.

- Security:
  - Sanitize `providerId` in `OidcAuthenticator` before logging to prevent log injection.
  - Validate `providerId` in `LoginController` to fix a trust boundary violation.

- Refactor:
  - Removed unused `providerId` and `request` parameters from the OIDC callback method.

- Feature:
  - Added an `iconUri` field to `OidcProviderConfig` to support custom provider icons on the login page.

- Configuration:
  - Updated `application.yaml` to include the new `icon-uri` property in the Keycloak provider example, demonstrating the feature.
  - Improved the commented-out `issuer-uri` example for better use in containerized environments (e.g., Docker).

tusaryan committed:
Enhance OIDC tests and handle null authorization URL

- Improves unit test coverage for the OIDC module by adding tests for failure scenarios, provider validation, and new fields like `iconUri`.
- Adds a null check in `LoginController` to gracefully redirect if the OIDC authorization URL fails to generate.

tusaryan committed:
Enhance OIDC tests and handle null authorization URL

- Improves unit test coverage for the OIDC module by adding tests for failure scenarios, provider validation, and new fields like `iconUri`.
- Adds a null check in `LoginController` to gracefully redirect if the OIDC authorization URL fails to generate.

tusaryan committed:
Enhance OIDC login API tests and docker setup

Expanded OIDC login API test coverage with additional negative and edge case scenarios, improved assertions, and refactored test structure. Updated OidcLoginPage to decouple from LoginPage and added helper methods for negative testing. Docker Compose and Keycloak realm export were updated for improved health checks, initialization, and OIDC configuration to support more robust integration testing.

tusaryan committed:
Enhance OIDC login API tests and docker setup

Expanded OIDC login API test coverage with additional negative and edge case scenarios, improved assertions, and refactored test structure. Updated OidcLoginPage to decouple from LoginPage and added helper methods for negative testing. Docker Compose and Keycloak realm export were updated for improved health checks, initialization, and OIDC configuration to support more robust integration testing.

SbloodyS commented:
Please merge these two PR #17410 and #17428 into this PR. @tusaryan

tusaryan committed:
Merge remote-tracking branch 'origin/feature/17171-oidc-ui' into feature/17171-oidc-core-backend

tusaryan committed:
Merge remote-tracking branch 'origin/feature/17171-oidc-api-test' into feature/17171-oidc-core-backend

SbloodyS commented on dolphinscheduler-api-test/dolphinscheduler-api-test-case/src/test/resources/docker/oidc-login/docker-compose.yaml:
```diff
@@ -0,0 +1,93 @@
+version: '3.8'
```
Missing license header.

SbloodyS commented on dolphinscheduler-api-test/dolphinscheduler-api-test-case/src/test/resources/docker/oidc-login/docker-compose.yaml:
```diff
@@ -0,0 +1,93 @@
+      interval: 5s
+      timeout: 10s
+      retries: 40
+      start_period: 180s
```
Please leave blank line at the end of each file.

tusaryan committed:
Add ASF license headers and MySQL initialization script for OIDC login Docker test environment

tusaryan committed:
Add ASF license headers and MySQL initialization script for OIDC login Docker test environment

Gallardot commented on dolphinscheduler-api-test/dolphinscheduler-api-test-case/src/test/resources/docker/oidc-login/init-sql/dolphinscheduler_mysql.sql:
```diff
@@ -0,0 +1,1370 @@
+-- ----------------------------
+-- Table structure for QRTZ_JOB_DETAILS
+-- ----------------------------
+DROP TABLE IF EXISTS `QRTZ_JOB_DETAILS`;
```
Is this file copy from 'https://github.com/apache/dolphinscheduler/blob/dev/dolphinscheduler-dao/src/main/resources/sql/dolphinscheduler_mysql.sql' ?

Would it be better to create a soft link? There is no need to maintain two files.

Gallardot commented on dolphinscheduler-api/src/main/resources/application.yaml:
```diff
@@ -216,6 +220,29 @@ security:
+        keycloak:
+          display-name: "Login with Keycloak" # Text for UI Button
+          issuer-uri: http://keycloak:8080/realms/dolphinscheduler
+          icon-uri: "/images/providers-icon/keycloak.png"
```
Just a example?

SbloodyS commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/LoginController.java:
```diff
@@ -263,4 +303,90 @@ public void loginByAuth2(@RequestParam String code, @RequestParam String provide
+     */
+    @SneakyThrows
+    @Operation(summary = "handleOidcCallback", description = "HANDLE_OIDC_CALLBACK")
+    @GetMapping("login/oauth2/code/{providerId}")
```
```suggestion
    @GetMapping("/login/oauth2/code/{providerId}")
```

SbloodyS commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/LoginController.java:
```diff
@@ -207,6 +222,31 @@ public Result<List<OAuth2Configuration.OAuth2ClientProperties>> oauth2Provider()
+     * @return list of OIDC providers
+     */
+    @Operation(summary = "getOidcProviders", description = "GET_OIDC_PROVIDERS")
+    @GetMapping("oidc-providers")
```
```suggestion
    @GetMapping("/oidc-providers")
```

SbloodyS commented on dolphinscheduler-ui/src/views/login/index.module.scss:
```diff
@@ -59,4 +59,41 @@
+  height: 20px;
+  margin-right: 8px;
+  object-fit: contain;
 }
```
Please leave an blank line at the bottom of each file.

SbloodyS commented on dolphinscheduler-api-test/dolphinscheduler-api-test-case/src/test/resources/docker/oidc-login/docker-compose.yaml:
```diff
@@ -0,0 +1,110 @@
+      - "12345:12345"
+    environment:
+      - TZ=Asia/Shanghai
+      - SPRING_PROFILES_ACTIVE=mysql
```
I think we don't need `mysql` since standalone server would use `h2` by default. The usage of mysql and h2 is basically the same. And we don't need init database sql file too.

tusaryan committed:
feat(api-test): Refactor OIDC test environment and improve test logic

Refactored the OIDC API test environment and the `OidcLoginAPITest` class to improve stability, precision, and alignment with project best practices.

Environment Changes:
- Updated `docker-compose.yaml` to use the `apache/dolphinscheduler-standalone-server:ci` image.
- Removed the external MySQL and Zookeeper services and the test now leverages the default H2 database and embedded Zookeeper from the standalone server, simplifying the setup and increasing execution speed.

Test Logic Improvements:
- Added `@Slf4j` for logging in `OidcLoginAPITest`.
- Made the assertion in `testLoginEndpointDisabledInOidcMode` more precise by checking for a specific "HTTP 401 Unauthorized" status code instead of a generic non-200 response.

tusaryan committed:
feat(api-test): Refactor OIDC test environment and improve test logic

Refactored the OIDC API test environment and the `OidcLoginAPITest` class to improve stability, precision, and alignment with project best practices.

Environment Changes:
- Updated `docker-compose.yaml` to use the `apache/dolphinscheduler-standalone-server:ci` image.
- Removed the external MySQL and Zookeeper services and the test now leverages the default H2 database and embedded Zookeeper from the standalone server, simplifying the setup and increasing execution speed.

Test Logic Improvements:
- Added `@Slf4j` for logging in `OidcLoginAPITest`.
- Made the assertion in `testLoginEndpointDisabledInOidcMode` more precise by checking for a specific "HTTP 401 Unauthorized" status code instead of a generic non-200 response.

tusaryan committed:
refactor(security): Improve OIDC implementation and resolve SonarQube issues

This commit includes a series of refactors and fixes to the OIDC authentication module in the API, as well as a formatting fix for the UI.

Backend (`dolphinscheduler-api`):
- LoginController:
  - Switched from field injection to constructor injection for all dependencies to align with modern Spring best practices and improve testability.
  - Added a leading slash to OIDC endpoint mappings (`/oidc-providers`, `/login/oauth2/code/{providerId}`) for consistency across the application.
  - Resolved a SonarQube warning by binding the `{providerId}` path variable in the OIDC callback handler and using it to validate the `state` parameter, enhancing security.

- OidcAuthenticator:
  - Eliminated duplicated string literals by introducing constants for the "email" attribute and the log sanitization regex, resolving high-impact SonarQube issues.

Frontend (`dolphinscheduler-ui`):
- Code Formatting:
  - Executed `pnpm run lint` to correct code style and formatting issues, ensuring the frontend build passes all CI checks.

tusaryan committed:
refactor(security): Improve OIDC implementation and resolve SonarQube issues

This commit includes a series of refactors and fixes to the OIDC authentication module in the API, as well as a formatting fix for the UI.

Backend (`dolphinscheduler-api`):
- LoginController:
  - Switched from field injection to constructor injection for all dependencies to align with modern Spring best practices and improve testability.
  - Added a leading slash to OIDC endpoint mappings (`/oidc-providers`, `/login/oauth2/code/{providerId}`) for consistency across the application.
  - Resolved a SonarQube warning by binding the `{providerId}` path variable in the OIDC callback handler and using it to validate the `state` parameter, enhancing security.

- OidcAuthenticator:
  - Eliminated duplicated string literals by introducing constants for the "email" attribute and the log sanitization regex, resolving high-impact SonarQube issues.

Frontend (`dolphinscheduler-ui`):
- Code Formatting:
  - Executed `pnpm run lint` to correct code style and formatting issues, ensuring the frontend build passes all CI checks.

tusaryan committed:
Enhance OIDC authentication error handling and config injection

Added specific OIDC error statuses to the Status enum and improved error handling in OidcAuthenticator for token exchange and ID token validation. Refactored SecurityConfig and OidcAuthenticator to use constructor-based dependency injection for OIDC configuration and UsersService, ensuring required beans are present for OIDC authentication.

tusaryan committed:
Enhance OIDC authentication error handling and config injection

Added specific OIDC error statuses to the Status enum and improved error handling in OidcAuthenticator for token exchange and ID token validation. Refactored SecurityConfig and OidcAuthenticator to use constructor-based dependency injection for OIDC configuration and UsersService, ensuring required beans are present for OIDC authentication.

tusaryan committed:
Add OIDC authentication docs and test setup guide

Expanded authentication documentation to include OIDC (OpenID Connect) support, detailing configuration steps for various providers and local development with Keycloak. Added a new guide for setting up a local OIDC environment and updated API test README with instructions for running OIDC-related API tests using Docker and Testcontainers.

tusaryan committed:
Add OIDC authentication docs and test setup guide

Expanded authentication documentation to include OIDC (OpenID Connect) support, detailing configuration steps for various providers and local development with Keycloak. Added a new guide for setting up a local OIDC environment and updated API test README with instructions for running OIDC-related API tests using Docker and Testcontainers.

Gallardot committed:
Merge branch 'dev' into feature/17171-oidc-core-backend

SbloodyS commented on dolphinscheduler-api-test/dolphinscheduler-api-test-case/src/test/java/org/apache/dolphinscheduler/api/test/cases/OidcLoginAPITest.java:
```diff
@@ -0,0 +1,117 @@
+import org.junit.jupiter.api.Order;
+import org.junit.jupiter.api.Test;
+
+@DolphinScheduler(composeFiles = "docker/oidc-login/docker-compose.yaml")
```
Please add this to https://github.com/apache/dolphinscheduler/blob/bccb8ec6ab4212f1dd80b9cb2b2d4c520cc93729/.github/workflows/api-test.yml#L120-L121

SbloodyS commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/LoginController.java:
```diff
@@ -82,17 +88,27 @@
-
-    @Autowired
-    private Authenticator authenticator;
+    private final SessionService sessionService;
+    private final Authenticator authenticator;
+    private final OAuth2Configuration oAuth2Configuration;
+    private final OidcConfigProperties oidcConfigProperties;
+    private final UsersService usersService;
```
Why change this?

SbloodyS commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/security/impl/oidc/OidcAuthenticator.java:
```diff
@@ -0,0 +1,390 @@
+    private final OidcConfigProperties oidcConfig;
+    private final UsersService usersService;
+
+    @Value("${api.base-url:http://localhost:12345/dolphinscheduler}")
+    private String apiBaseUrl;
```
There are multiple parameters like this, and we should merge them into one.

ruanwenjun replied:
+1

SbloodyS commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/security/impl/oidc/OidcAuthenticator.java:
```diff
@@ -0,0 +1,390 @@
+                return null;
+            }
+
+            String[] stateParts = state.split(":", 2);
```
Please using Constants instead.

SbloodyS commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/security/impl/oidc/OidcAuthenticator.java:
```diff
@@ -0,0 +1,390 @@
+
+            String[] stateParts = state.split(":", 2);
+            if (stateParts.length != 2) {
+                String sanitizedState = state.replaceAll(SANITIZE_REGEX, "_");
```
Same here.

SbloodyS commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/security/impl/oidc/OidcAuthenticator.java:
```diff
@@ -0,0 +1,390 @@
+            String providerId = stateParts[0];
+            OidcProviderConfig providerConfig = oidcConfig.getProviders().get(providerId);
+            if (providerConfig == null) {
+                String sanitizedProviderId = providerId.replaceAll(SANITIZE_REGEX, "_");
```
Same here.

SbloodyS commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/security/impl/oidc/OidcAuthenticator.java:
```diff
@@ -0,0 +1,390 @@
+
+            return user;
+        } catch (Exception e) {
+            log.error("An error occurred during OIDC authentication.", e);
```
```suggestion
            log.error("An error occurred during OIDC authentication: ", e);
```

SbloodyS commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/security/impl/oidc/OidcAuthenticator.java:
```diff
@@ -0,0 +1,390 @@
+    public String getSignInUrl(String state) {
+
+        try {
+            String[] stateParts = state.split(":", 2);
```
Same here.

SbloodyS commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/security/impl/oidc/OidcAuthenticator.java:
```diff
@@ -0,0 +1,390 @@
+
+            return authRequest.toURI().toString();
+        } catch (Exception e) {
+            log.error("Error generating OIDC sign-in URL", e);
```
```suggestion
            log.error("Error generating OIDC sign-in URL: ", e);
```

SbloodyS commented on dolphinscheduler-api-test/dolphinscheduler-api-test-case/src/test/java/org/apache/dolphinscheduler/api/test/entity/HttpResponse.java:
```diff
@@ -29,4 +32,12 @@ public class HttpResponse {
+
+    private Map<String, String> headers = new HashMap<>();
+
+    public HttpResponse(int statusCode, HttpResponseBody body) {
+        this.statusCode = statusCode;
+        this.body = body;
+        this.headers = new HashMap<>();
+    }
```
Why we need this sinice we already hava `@AllArgsConstructor` and `private Map<String, String> headers = new HashMap<>();`

ruanwenjun replied:
It would be better to add `@Builder` if we hope to add many constructors.

ruanwenjun commented on docs/docs/en/guide/security/oidc-local-environment-setup.md:
```diff
@@ -0,0 +1,111 @@
+                auto-create: true
+                default-tenant-code: "default"
+                default-queue: "default"
+                admin-group-mapping:
```
```suggestion
                user-type: "ADMIN_USER"
```

tusaryan committed:
Refactor OIDC config and logging, add OIDC API test

Refactored OIDC authentication to use ApiConfig for base and UI URLs, improving configuration consistency. Enhanced logging by sanitizing sensitive state and provider IDs. Updated documentation for OIDC setup, removed unused constructor in HttpResponse, and added OidcLoginAPITest to the API test workflow.

tusaryan committed:
Refactor OIDC config and logging, add OIDC API test

Refactored OIDC authentication to use ApiConfig for base and UI URLs, improving configuration consistency. Enhanced logging by sanitizing sensitive state and provider IDs. Updated documentation for OIDC setup, removed unused constructor in HttpResponse, and added OidcLoginAPITest to the API test workflow.

tusaryan committed:
Merge branch 'feature/17171-oidc-core-backend' of https://github.com/tusaryan/dolphinscheduler into feature/17171-oidc-core-backend

SbloodyS committed:
Merge branch 'dev' into feature/17171-oidc-core-backend

tusaryan committed:
Fix OIDC provider null handling and update tests

Adds a null check for missing OIDC providers in OidcAuthenticator to prevent errors. Updates related tests to mock ApiConfig instead of using a @Value property, and adjusts test setup accordingly. Also updates OidcLoginPage to use an empty HashMap for request parameters.

tusaryan committed:
Fix OIDC provider null handling and update tests

Adds a null check for missing OIDC providers in OidcAuthenticator to prevent errors. Updates related tests to mock ApiConfig instead of using a @Value property, and adjusts test setup accordingly. Also updates OidcLoginPage to use an empty HashMap for request parameters.

tusaryan commented:
Hi @SbloodyS , @Gallardot, @ruanwenjun when you have a moment, could you help me troubleshoot some failing API tests? Thank you!

SbloodyS commented:
> Hi @SbloodyS , @Gallardot, @ruanwenjun when you have a moment, could you help me troubleshoot some failing API tests? Thank you!

OIDCLoginAPITest failed due to container did not start healthy. For more info you can take at look at https://github.com/apache/dolphinscheduler/actions/runs/17434069882/job/49556012570?pr=17340

ruanwenjun commented:
Can it safe to place this in 3.3.2?

tusaryan committed:
Merge branch 'feature/17171-oidc-core-backend' of https://github.com/tusaryan/dolphinscheduler into feature/17171-oidc-core-backend

tusaryan committed:
feat(api): Implement robust OIDC role synchronization and API tests fix

This commit resolves a proper role synchronization logic and adds a comprehensive API test suite to validate the entire OIDC feature.
The solution involved these changes:

1.  Implement Role Synchronization on Every Login:
    - The logic in `OidcAuthenticator` has been refactored to check and update a user's role on every login, not just on initial creation. This ensures the OIDC provider remains the single source of truth for user roles.
    - A new, simplified `updateUser(User user)` method was added to the `UsersService` layer to support this synchronization in a clean, decoupled manner.

2.  Fix Configuration Property Loading:
    - A subtle but critical bug was fixed where nested OIDC configuration properties (specifically security.authentication.oidc.user.admin-group-mapping) were not being loaded from application.yaml.
    - This was resolved by adding the `@NestedConfigurationProperty` annotation in `OidcConfigProperties`, which is the standard Spring Boot practice for ensuring nested configuration objects are correctly populated.

3.  Improved OIDC API Test Suite:
    - The test suite provides coverage for:
        - Correct redirect behavior for valid and invalid providers.
        - Graceful handling of OIDC callback errors and malformed requests.
        - Security policy enforcement (verifying password login is disabled in OIDC mode).
        - A full, successful login simulation for a general user to ensure no regressions.
4. Other improvements include:
    - Replaced the sanitizeForLogging method in OidcAuthenticator with the existing RegexUtils.escapeNRT method to promote code reuse and remove redundant code.
    - API test cleanup: Introduce enums to replace scattered hardcoded delimiters and HTML entity constants. Apply replacements only where they improved clarity and maintainability.

tusaryan committed:
feat(api): Implement OIDC role sync and enhance API tests

This commit resolves a critical issue where OIDC user roles were not correctly assigned or updated. It introduces robust role synchronization on every login, adds a test to check Admin User Successfull Login in API test, and improves the developer documentation.
Change Log:
1.  Fix Role Synchronization bug in OidcAuthenticator:
The logic in `OidcAuthenticator` has been refactored to check and update a user's role on every login, not just on initial creation. This ensures the OIDC provider remains the single source of truth for user roles and closes a security vulnerability where role de-provisioning was ignored.

2.  Enhance OIDC API Test to coverage a full, successful login simulation for admin users.

3.  Updated the development documentation to use the Maven Wrapper (`./mvnw`) instead of the local `mvn` command. This ensures all contributors use the same Maven version as the project, reducing version compatibility issues and developer setup overhead.

tusaryan committed:
Fix OIDC authenticator test and improve coverage

tusaryan committed:
test(api): Improve unit test code coverage and apply code styles

Gallardot committed:
Merge branch 'dev' into feature/17171-oidc-core-backend

SbloodyS reviewed (approved):
LGTM

ruanwenjun reviewed (approved):
LGTM
